examples/amd-azure/multi_process_benchmark.py

```diff
@@ -0,0 +1,300 @@
+
+
+def decide_thread_pinning(pinning_mode: str):
+    pinning_mode = pinning_mode.lower()
+
+    if pinning_mode in "core":
+        os.environ["NM_BIND_THREADS_TO_CORES"] = "1"
+        _LOGGER.info("Thread pinning to cores enabled")
+    elif pinning_mode in "numa":
+        os.environ["NM_BIND_THREADS_TO_CORES"] = "0"
+        os.environ["NM_BIND_THREADS_TO_SOCKETS"] = "1"
+        _LOGGER.info("Thread pinning to socket/numa nodes enabled")
+    elif pinning_mode in "none":
+        os.environ["NM_BIND_THREADS_TO_CORES"] = "0"
+        os.environ["NM_BIND_THREADS_TO_SOCKETS"] = "0"
+        _LOGGER.info("Thread pinning disabled, performance may be sub-optimal")
+    else:
+        _LOGGER.info(
+            "Recieved invalid option for thread_pinning '{}', skipping".format(
+                pinning_mode
+            )
+        )
+
+
+def cpus_to_numas(cpu_set):
+    numas = set() 
+    for i in range(numa.info.get_max_node()):
+        cpus = numa.info.node_to_cpus(i)
+        for j in cpus:
+            if j in cpu_set:
+                numas.add(i)
+                break
+    return numas
+
+
+def run(worker_id, args, barrier, cpu_affinity_set, results):
+    # Set the CPU affinity of this process
+    numas = cpus_to_numas(cpu_affinity_set)
+    numa.memory.set_interleave_nodes(*numas)
+    numa.schedule.run_on_nodes(*numas)
+
+    # Run on given CPUs. The first argument is the process id,
+    # and 0 means this present process
+    numa.schedule.run_on_cpus(0, *cpu_affinity_set)
+
+    input_shapes = parse_input_shapes(args.input_shapes)
+
+    # Compile the model
+    model = compile_model(
+        model=args.model_path,
+        batch_size=args.batch_size,
+        num_cores=args.num_cores,
+        scheduler="single_stream",
+        input_shapes=input_shapes,
+    )
+
+    # Cleanly separate compilation and benchmarking
+    barrier.wait()
+
+    # Generate random inputs to feed the model
+    # TODO(mgoin): should be able to query Engine class instead of loading ONNX
+    if input_shapes:
+        with override_onnx_input_shapes(args.model_path, input_shapes) as model_path:
+            input_list = generate_random_inputs(model_path, args.batch_size)
+    else:
+        input_list = generate_random_inputs(args.model_path, args.batch_size)
+
+    # Warmup the engine
+    singlestream_benchmark(model, input_list, args.warmup_time)
+
+    # Run the benchmark scenario and collect batch times
+    batch_times = singlestream_benchmark(model, input_list, args.time)
+
+    # Cooldown the engine
+    singlestream_benchmark(model, input_list, args.warmup_time)
+
+    if len(batch_times) == 0:
+        raise Exception(
+            "Generated no batch timings, try extending benchmark time with '--time'"
+        )
+
+    results[worker_id] = batch_times
+
+
+def main():
+    args = parse_args()
+
+    decide_thread_pinning(args.thread_pinning)
+
+    b = mp.Barrier(args.num_streams)
+
+    # Hardcode affinity sets for elmo
```

Comment:
Change to Azure machine?

Reply:
I plan to move this into a file and pass it in as a parameter, and provide examples for both our machine and the Azure instance as well.